extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -378,37 +430,50 @@ public boolean apply(Throwable input)
               }
             }
 
-            final long nextOffset = record.offset() + 1;
-            final long endOffset = ioConfig.getEndPartitions().getPartitionOffsetMap().get(record.partition());
-
-            nextOffsets.put(record.partition(), nextOffset);
+            nextOffsets.put(record.partition(), record.offset() + 1);
+          }
 
-            if (nextOffset == endOffset && assignment.remove(record.partition())) {
-              log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
-              assignPartitions(consumer, topic, assignment);
-              stillReading = !assignment.isEmpty();
-            }
+          if (nextOffsets.get(record.partition()).equals(endOffsets.get(record.partition()))
+              && assignment.remove(record.partition())) {
+            log.info("Finished reading topic[%s], partition[%,d].", record.topic(), record.partition());
+            assignPartitions(consumer, topic, assignment);
+            stillReading = ioConfig.isPauseAfterRead() || !assignment.isEmpty();
           }
         }
       }
 
-      // Persist pending data.
-      final Committer finalCommitter = committerSupplier.get();
-      driver.persist(finalCommitter);
+      // if we didn't actually generate any segments, don't try to publish as this will throw errors on the overlord
+      if (driver.getActiveSegmentCount() == 0) {
+        log.info("No segments were created by this task");
+        stopGracefully();
+      } else {
+        // Persist pending data.
+        driver.persist(committerSupplier.get());
+      }
 
-      publishing = true;
-      if (stopping) {
-        // Stopped gracefully. Exit code shouldn't matter, so fail to be on the safe side.
-        return TaskStatus.failure(getId());
+      if (stopRequested && !publishOnStop) {
+        throw new InterruptedException("stopping without publishing");
       }
 
+      status = Status.PUBLISHING;
       final TransactionalSegmentPublisher publisher = new TransactionalSegmentPublisher()
       {
         @Override
         public boolean publishSegments(Set<DataSegment> segments, Object commitMetadata) throws IOException
         {
-          // Sanity check, we should only be publishing things that match our desired end state.
-          if (!ioConfig.getEndPartitions().equals(((Map) commitMetadata).get(METADATA_NEXT_PARTITIONS))) {
+          final KafkaPartitions finalPartitions = toolbox.getObjectMapper().convertValue(
+              ((Map) commitMetadata).get(METADATA_NEXT_PARTITIONS),
+              KafkaPartitions.class
+          );
+
+          if (stopRequested) {
```

Comment:
When stopGracefullyAndPublish() is called, `finalPartitions.getPartitionOffsetMap` will not necessarily have reached `endOffsets`, but will stop and publish whatever it currently has. I suppose we could also just set `endOffsets` to `finalPartitions.getPartitionOffsetMap` if we've been signaled to stop, but logically the end result would be pretty similar

(To note, I'm not using stopGracefullyAndPublish() right now in the supervisor because I wanted to be able to synchronize replica tasks to ensure we've read the exact same partitions/offsets before publishing so I'm using the pause-then-setEndOffset mechanism)